ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineContext.java

```diff
@@ -208,16 +231,90 @@ DdlCommand createDdlCommand(
   String executeDdl(
       final String sqlExpression,
       final DdlCommand command,
-      final boolean withQuery
+      final boolean withQuery,
+      final Set<SourceName> withQuerySources
   ) {
-    final DdlCommandResult result = ddlCommandExec.execute(sqlExpression, command, withQuery);
+    if (command instanceof DropSourceCommand) {
+      throwIfOtherQueriesExist(((DropSourceCommand) command).getSourceName());
+    }
+
+    final DdlCommandResult result =
+        ddlCommandExec.execute(sqlExpression, command, withQuery, withQuerySources);
     if (!result.isSuccess()) {
       throw new KsqlStatementException(result.getMessage(), sqlExpression);
     }
+
+    if (command instanceof DropSourceCommand) {
+      terminateCreateAsQuery(((DropSourceCommand) command).getSourceName());
+    }
+
     return result.getMessage();
   }
 
-  void registerQuery(final QueryMetadata query) {
+  private void terminateCreateAsQuery(final SourceName sourceName) {
+    createAsQueries.computeIfPresent(sourceName, (ignore , queryId) -> {
+      final PersistentQueryMetadata query = persistentQueries.get(queryId);
+      if (query != null) {
+        query.close();
+      }
+
+      return null;
+    });
+  }
+
+  private Set<String> getOtherQueriesWithSink(final SourceName sourceName) {
+    final ImmutableSet.Builder<String> queries = ImmutableSet.builder();
+
+    if (otherQueries.containsKey(sourceName)) {
+      otherQueries.get(sourceName).forEach(queryId -> {
+        final PersistentQueryMetadata query = persistentQueries.get(queryId);
+        if (query != null && query.getSinkName().equals(sourceName)) {
```

Comment:
if this condition is ever not met, don't we have some referential integrity issue? if we added a query to the `otherQueries` under the `sourceName` key, shouldn't it always be the case that there exists a query and the sink is `sourceName`? same question goes for `getOtherQueriesWithSource`

Reply:
The query that is referenced is not always a sink. See this example:

```
// insert1  
// query: INSERT INTO s1 SELECT FROM s0; 
// sink: s1, sources: s0
otherQueries
- s1 -> [insert1]
- s0 -> [insert1]

  
// insert2 
// query: INSERT INTO s2 SELECT FROM s1; 
// sink: s2, sources: s1
otherQueries
- s2 -> [insert2]
- s1 -> [insert1, insert2]
- s0 -> [insert1]
```

The queries with sink for `s1` should only be `insert1`, but not `insert2`. So I need to get that list and check that the sink for those queries matches the source name I'm looking at (same for source for those queries).

I initially thought of using two maps, otherQueriesSources and otherQueriesSinks. But I was just adding too much stuff in the code, and I saw this improvement of having only one. We don't have too many queries to cause a performance penalty walking through the queries of the desired source.

I made a small change to see this filter better, like:
```
final Set<QueryId> sinkQueries = getOtherQueries(sourceName, FILTER_QUERIES_WITH_SINK);
final Set<QueryId> sourceQueries = getOtherQueries(sourceName, FILTER_QUERIES_WITH_SOURCE);
```